ENH: Modulate dispatched x86 CPU features

seiko2plus opened the pull request:
## Overview

This PR reorganizes NumPy's CPU build options by replacing individual x86 features with microarchitecture levels. This change aligns with the Google Highway project requirements and common Linux distribution practices.

This PR default setting for `cpu-baseline` on x86 has been raised to `x86-64-v2` microarchitecture as we're in 2025 and adding SIMD compatibility for antiquated  CPUs from before 2009 is no longer practical or efficient.
This can be changed to `cpu-baseline=none` during build time to support older CPUs, though manual SIMD optimizations for pre-2009 processors are no longer supported. This change improves performance and reduces binary 
size while only affecting hardware that is over 15 years old. 

## Key Changes

- **Consolidated into Microarchitecture Groups**: Replaced individual features with `X86_V2`, `x86_V3`, and `x86_V4` groups
- **Adjusted Baseline**: Set to micro-architecture level 2 (`x86_V2`), covering features from CPUs since 2009.
  This improves performance and reduces binary size
- **Improved `-` Operator**: Corrected to properly exclude successor features
- **Backward Compatibility**: Added redirection via meson for removed feature names
- **Stricter Compatibility**: Features like AVX512 without full mask operations now considered unsupported rather than using fallbacks

## Detailed CPU Feature Changes

- Removed individual features (`SSE`, `SSE2`, `SSE3`, `SSSE3`, `SSE4_1`, `SSE4_2`, `POPCNT`) → now in `X86_V2`
- Removed AMD legacy features (`XOP`, `FMA4`)
- Removed Xeon Phi support (`AVX512_KNL`, `AVX512_KNM`)
- Removed individual features (`AVX`, `AVX2`, `FMA3`, `F16C`) → now in `X86_V3`
- Removed `AVX512F`, `AVX512CD` (from dropping Xeon Phi support)
- Renamed `AVX512_SKX` to `X86_V4`
- Removed redundant groups `AVX512_CLX` and `AVX512_CNL`
- Updated `AVX512_ICL` to include `VAES`, `GFNI`, `VPCLMULQDQ`

## New Feature Group Hierarchy

| Name | Implies | Includes |
|------|---------|----------|
| X86_V2 | | SSE SSE2 SSE3 SSSE3 SSE4_1 SSE4_2 POPCNT CX16 LAHF |
| X86_V3 | X86_V2 | AVX AVX2 FMA3 BMI BMI2 LZCNT F16C MOVBE |
| X86_V4 | X86_V3 | AVX512F AVX512CD AVX512VL AVX512BW AVX512DQ |
| AVX512_ICL | X86_V4 | AVX512VBMI AVX512VBMI2 AVX512VNNI AVX512BITALG AVX512VPOPCNTDQ AVX512IFMA VAES GFNI VPCLMULQDQ |
| AVX512_SPR | AVX512_ICL | AVX512FP16 |

## CPU Generation Mapping

- `X86_V2`: x86-64-v2 microarchitectures (CPUs since 2009)
- `X86_V3`: x86-64-v3 microarchitectures (CPUs since 2015)
- `X86_V4`: x86-64-v4 microarchitectures (AVX-512 capable CPUs)
- `AVX512_ICL`: Intel Ice Lake and similar CPUs
- `AVX512_SPR`: Intel Sapphire Rapids and newer CPUs

> Note: On 32-bit x86, `cx16` is excluded from `X86_V2`.

## Documentation

Documentation has been updated to reflect these changes and to fit the current meson build system.

closes #27851

seiko2plus committed:
ENH: Modulate dispatched x86 CPU features

**IMPORTANT NOTE**: The default setting for `cpu-baseline`` on x86 has been raised to `x86-64-v2` microarchitecture.
This can be changed to `cpu-baseline=none` during build time to support older CPUs, though manual SIMD optimizations for pre-2009 processors are no longer supported.

This patch reorganizes CPU build options by replacing individual x86 features with microarchitecture levels.
This change aligns with the Google Highway requirements and common Linux distribution practices.

This patch:
- Removes all individual x86 features and replaces them with three microarchitecture
  levels (`X86_V2`, `X86_V3`, `X86_V4`) commonly used by Linux distributions
- Raises the baseline to microarchitecture level 2 (replacing `SSE3`) since all known
  x86 CPUs since 2009 support it. This improves performance and
  reduces binary size
- Updates documentation to to reflect these changes and to fit the current meson build system.
- Corrects the behavior of the `-` operator, which now excludes successor features
  that imply the excluded feature
- Adds redirection via meson for removed feature names to avoid breaking builds
- Removes compiler compatibility workarounds, so features like AVX512 without full mask operations
  will be considered unsupported rather than providing fallbacks

Detailed CPU features changes:
- Removes individual features (`SSE`, `SSE2`, `SSE3`, `SSSE3`, `SSE4_1`, `SSE4_2`,
  `POPCNT`) which are now part of the new group `X86_V2`
- Removes AMD legacy features (`XOP`, `FMA4`)
- Removes Xeon Phi support (`AVX512_KNL`, `AVX512_KNM`) which Intel has discontinued
- Removes individual features (`AVX`, `AVX2`, `FMA3`, `F16C`) which are now part
  of the new group `X86_V3`
- Removes individual features `AVX512F`, `AVX512CD` as a result of dropping Xeon Phi support
- Renames group `AVX512_SKX` to `x86_v4` to align with microarchitecture level naming
- Removes groups `AVX512_CLX` and `AVX512_CNL` (features available via `AVX512_ICL`)
- Updates `AVX512_ICL` to include features (`VAES`, `GFNI`, `VPCLMULQDQ`) for alignment
  with Highway

New Feature Group Hierarchy:
```
Name          | Implies     | Includes
--------------|-------------|-----------------------------------------------------------
X86_V2        |             | SSE SSE2 SSE3 SSSE3 SSE4_1 SSE4_2 POPCNT CX16 LAHF
X86_V3        | X86_V2      | AVX AVX2 FMA3 BMI BMI2 LZCNT F16C MOVBE
X86_V4        | X86_V3      | AVX512F AVX512CD AVX512VL AVX512BW AVX512DQ
AVX512_ICL    | X86_V4      | AVX512VBMI AVX512VBMI2 AVX512VNNI AVX512BITALG
              |             | AVX512VPOPCNTDQ AVX512IFMA VAES GFNI VPCLMULQDQ
AVX512_SPR    | AVX512_ICL  | AVX512FP16
```

These groups correspond to CPU generations:
- `X86_V2`: x86-64-v2 microarchitectures (CPUs since 2009)
- `X86_V3`: x86-64-v3 microarchitectures (CPUs since 2015)
- `X86_V4`: x86-64-v4 microarchitectures (AVX-512 capable CPUs)
- `AVX512_ICL`: Intel Ice Lake and similar CPUs
- `AVX512_SPR`: Intel Sapphire Rapids and newer CPUs

Note: On 32-bit x86, `cx16` is excluded from `X86_V2`.

seiko2plus committed:
Fix: Clear FP status to prevent invalid left_shift warnings with `x86-64-v2` baseline

Prevents "invalid value encountered in left_shift" warnings on clang-cl
when testing bit shifts with long types under `x86-64-v2` baseline.

seiko2plus committed:
Set `-mfpmath=sse` on `x86` for gcc/clang numeric consistency

Force SSE-based floating-point on 32-bit x86 systems to fix inconsistent
results between einsum and other math functions. Prevents test failures
with int16 operations by avoiding the x87 FPU's extended precision.

jorenham commented:
needs a rebase

tacaswell commented:
If I am understanding 

>  build from source and change the default baseline to cpu-baseline=none.

correctly, than the limitation on older machines depends on how the binaries are built?  

For any of the downstream packagers (linux distros, conda-forge, etc) that is their responsibilities to sort out. For wheels I think there is a case to go more aggressively newer and push users towards better packaging ecosystems if they need to support older chips.

seiko2plus commented:
You're right -  we're bumping the default `cpu-baseline` and users can change this. I agree. Downstream packagers need to decide their own compatibility targets but we will no longer need to provide SIMD kernels below `x86-64-v2`. See the updated documentation for clarification:

https://github.com/numpy/numpy/blob/729d61c84d087dccf17e4ef960391227aaa0c5b3/doc/source/reference/simd/build-options.rst#targeting-older-cpus

seiko2plus commented:
@tacaswell, I've updated the release note to be more clear about the CPU baseline change.

seiko2plus commented on numpy/_core/meson.build:
```diff
@@ -1051,7 +1051,7 @@ foreach gen_mtargets : [
     src_file.process('src/umath/loops_unary_complex.dispatch.c.src'),
     [
-      AVX512F, [AVX2, FMA3], SSE2,
+      X86_V3, X86_V2,
```
```suggestion
      X86_V4, X86_V3, X86_V2,
```
AVX512 is missing here.

seiko2plus commented on numpy/_core/src/common/npy_cpu_features.c:
```diff
@@ -504,58 +528,97 @@ npy__cpu_init_features(void)
+
+
+    // Legacy groups
+    npy__cpu_have[NPY_CPU_FEATURE_AVX512_KNL] = npy__cpu_have[NPY_CPU_FEATURE_AVX512F] &&
+                                                npy__cpu_have[NPY_CPU_FEATURE_AVX512CD] &&
+                                                npy__cpu_have[NPY_CPU_FEATURE_AVX512ER] &&
+                                                npy__cpu_have[NPY_CPU_FEATURE_AVX512PF];
+
+    npy__cpu_have[NPY_CPU_FEATURE_AVX512_KNM] = npy__cpu_have[NPY_CPU_FEATURE_AVX512_KNL] &&
+                                                npy__cpu_have[NPY_CPU_FEATURE_AVX5124FMAPS] &&
+                                                npy__cpu_have[NPY_CPU_FEATURE_AVX5124VNNIW] &&
+                                                npy__cpu_have[NPY_CPU_FEATURE_AVX512VPOPCNTDQ];
+
+    npy__cpu_have[NPY_CPU_FEATURE_AVX512_CLX] = npy__cpu_have[NPY_CPU_FEATURE_X86_V4] &&
+                                                npy__cpu_have[NPY_CPU_FEATURE_AVX512VNNI];
+
+    npy__cpu_have[NPY_CPU_FEATURE_AVX512_CNL] = npy__cpu_have[NPY_CPU_FEATURE_X86_V4] &&
+                                                npy__cpu_have[NPY_CPU_FEATURE_AVX512IFMA] &&
+                                                npy__cpu_have[NPY_CPU_FEATURE_AVX512VBMI];
```
```suggestion
```
Not sure if we should drop legacy groups. I kept them in case there are projects that depend on them via dict `__cpu_features__`

seiko2plus replied:
sounds a good call to me

seiko2plus commented on .github/workflows/linux_simd.yml:
```diff
@@ -162,12 +162,12 @@ jobs:
           - [
             "without avx512/avx2/fma3",
-            "-Dallow-noblas=true -Dcpu-dispatch=SSSE3,SSE41,POPCNT,SSE42,AVX,F16C",
+            "-Dallow-noblas=true -Dcpu-dispatch=max-x86_v3",
```
```suggestion
            "-Dallow-noblas=true -Dcpu-dispatch=max-x86_v3",
```
This test is no longer needed, since there's a CI with baseline only.

seiko2plus commented on doc/source/reference/simd/build-options.rst:
```diff
@@ -1,222 +1,376 @@
-     are not supported by the target CPU (raises Python runtime error).
+- ``cpu-baseline``: The minimum set of CPU features required to run the compiled NumPy.
+  
+  * Default: ``min`` (provides compatibility across a wide range of platforms)
```
```suggestion
  * Default: ``min`` (provides compatibility across a wide range of platforms), see :ref:`special options <opt-special-options>` to check which min maps to for each architecture.
```
provides a ref for option `min`

seiko2plus commented on meson_cpu/meson.build:
```diff
@@ -207,8 +226,17 @@ foreach opt_name, conf : parse_options
     else
       filterd = []
+      # filter out the features that are in the accumulate list 
+      # including any successor features
```
Not excluding successor features should be considered a bug and needs to be back-ported.
For example, with cpu-dispatch before this patch, to exclude AVX512 you had to exclude all successor features, otherwise it would not be disabled:
```
python -m build --wheel -Csetup-args=-Dcpu-dispatch="max -avx512f -avx512cd \
-avx512_knl -avx512_knm -avx512_skx -avx512_clx -avx512_cnl -avx512_icl -avx512_spr"
 ```
After this fix:
```
python -m build --wheel -Csetup-args=-Dcpu-dispatch="max -avx512f"
```

seiko2plus commented on meson_cpu/x86/meson.build:
```diff
@@ -1,228 +1,102 @@
-  test_code: files(source_root + '/numpy/distutils/checks/cpu_avx512_cnl.c')[0]
-)
+HWY_SSE4_FLAGS = ['-DHWY_WANT_SSE4', '-DHWY_DISABLE_PCLMUL_AES']
+# Use SSE for floating-point on x86-32 to ensure numeric consistency.
```
This fix needs to be backported too.

seiko2plus replied:
It will be backported in #30026, and yes, I wonder if you have encountered such ambiguous runtime floating-point errors on 32-bit builds?

seiko2plus commented on meson_cpu/x86/meson.build:
```diff
@@ -1,228 +1,102 @@
+  test_code: files(current_dir + '/test_x86_v4.c')[0],
+)
+if cpu_family == 'x86'
+  X86_V4.update(disable: 'not supported on x86-32')
```
This disable needs to be backported too. We should not generate `AVX512` kernels on 32-bit systems.

seiko2plus commented on meson_cpu/x86/meson.build:
```diff
@@ -1,228 +1,102 @@
-  AVX2.update(args: {'val': '/arch:AVX2', 'match': clear_arch})
-  AVX512_SKX.update(args: {'val': '/arch:AVX512', 'match': clear_arch})
+  X86_V3.update(args: {'val': '/arch:AVX2', 'match': clear_arch})
+  X86_V4.update(args: {'val': '/arch:AVX512', 'match': clear_arch})
```
By default, Highway considers AVX512 a broken platform on MSVC. We've managed to support AVX512 on MSVC through universal intrinsics. We have three options here:

1. Keep it as-is: This leads to building AVX2 kernels with `/arch:AVX512`, which gives the compiler the opportunity to optimize (hopefully). Worst case: we'll have duplicate kernels.

2. Disable AVX512 support entirely:
```suggestion
  X86_V4.update(disable: 'not supported by Highway')
```
3.Change the default behavior and force-enable AVX512 on MSVC and lets see how deep the rabbit-hole goes:
```suggestion
  X86_V4.update(args: [{'val': '/arch:AVX512', 'match': clear_arch}, '-DHWY_BROKEN_MSVC=0'])
```

rgommers replied:
I think I'd prefer (2) or (3). If it's really very broken, disabling it seems fine (and low-effort). If it's feasible and there is energy for this, fixing it up in Highway over time seems nice. 

(1) seems worst, if it doesn't do much, I'd much rather take the binary size gains from disabling it completely.

seiko2plus replied:
We still haven't converted many kernels yet from universal intrinsics to Highway, so adding extra burden isn't a wise decision. Let's go for option (2) for now and investigate later which MSVC versions are compatible with Highway.

seiko2plus reviewed (commented):
Three cases require changes; the rest are backport notes and comments that may need your decision.

seiko2plus committed:
Address review feedback on CPU features and CI workflows

- Disable X86_V4 (AVX-512) for MSVC builds due to Highway incompatibility
- Add FIXME comment for future MSVC compatibility investigation
- Update SIMD CI workflow to reflect `x86-64-v2` baseline
- Remove redundant test configurations
- Add missing X86_V4 support to unary complex loops

r-devulap commented on doc/release/upcoming_changes/28896.change.rst:
```diff
@@ -0,0 +1,56 @@
+
+These groups correspond to CPU generations:
+
+- ``X86_V2``: x86-64-v2 microarchitectures (CPUs since 2009)
```
The names X86_V2, X86_V3, and X86_V4 are ambiguous and it is not immediately obvious what they represent. I think we can do a better job naming them to something like: SSE42, AVX2_FMA, AVX512_SKX?

r-devulap commented on meson_cpu/x86/test_x86_v4.c:
```diff
@@ -0,0 +1,88 @@
+    __m512i avx512dq_i64_a = _mm512_set1_epi64(seed);
+    __m512i avx512dq_i64_b = _mm512_set1_epi64(2);
+    __m512i avx512dq_i64_c = _mm512_add_epi64(avx512dq_i64_a, avx512dq_i64_b);
+    result += _mm_cvtsi128_si32(_mm512_extracti32x4_epi32(avx512dq_i64_c, 0));
```
This seem unnecessarily too long. We should be able to keep this down to 5 or 6 instructions., something that tests one intrinsic from each feature: 

```
    // Test one intrinsic from each extension
    
    // AVX-512F (Foundation)
    __m512 avx512f_val = _mm512_set1_ps(1.0f);
    
    // AVX-512CD (Conflict Detection)
    __m512i avx512cd_val = _mm512_conflict_epi32(_mm512_set1_epi32(1));
    
    // AVX-512VL (Vector Length Extensions)
    __m256i avx512vl_val = _mm256_mask_mov_epi32(_mm256_set1_epi32(0), 0xFF, _mm256_set1_epi32(1));
    
    // AVX-512BW (Byte and Word)
    __m512i avx512bw_val = _mm512_set1_epi8(1);
    
    // AVX-512DQ (Doubleword and Quadword)
    __m512d avx512dq_val = _mm512_set1_pd(1.0);
    
    // Prevent unused variable warnings
    (void)avx512f_val;
    (void)avx512cd_val;
    (void)avx512vl_val;
    (void)avx512bw_val;
    (void)avx512dq_val;

```

seiko2plus replied:
Each line ensures both the compiler and assembler fully support all AVX-512 feature groups. Mask operations make the test longer, but they’re now required: Highway by default falls back to AVX2 if full AVX-512 isn’t fully supported they marks these compilers as broken. Before this patch, partial AVX-512/BW/DQ support is possible as universal intrinics has special workarounds for these cases you can follow `AVX512BW_MASK` and `AVX512DQ_MASK` for more details. However this compile-time test drop this support, and I have left a release note for it.

seiko2plus replied:
Another reason the test is longer is the inclusion of shuffle and scalar extraction operations, which are important to prevent the compiler from optimizing them out and to ensure proper assembler support.

r-devulap commented on meson_cpu/x86/test_x86_v3.c:
```diff
@@ -0,0 +1,66 @@
+    __m128 f16c_src = _mm_set1_ps((float)seed);
+    __m128i f16c_half = _mm_cvtps_ph(f16c_src, 0);
+    __m128 f16c_restored = _mm_cvtph_ps(f16c_half);
+    result += (int)_mm_cvtss_f32(f16c_restored);
```
Same comment as below, can we keep this simple please?

r-devulap reviewed (changes requested):
Completed an initial pass; detailed review still pending.

seiko2plus committed:
fix up msvc

seiko2plus commented on doc/release/upcoming_changes/28896.change.rst:
```diff
@@ -0,0 +1,56 @@
+     - Includes
+   * - ``X86_V2``
+     - 
+     - ``SSE`` ``SSE2`` ``SSE3`` ``SSSE3`` ``SSE4_1`` ``SSE4_2`` ``POPCNT`` ``CX16`` ``LAHF``
```
By default, Highway’s SSE4 isn’t compatible with X86_V2 (Microarchitecture Level 2). We need to pass `-DHWY_DISABLE_PCLMUL_AES` to enable it.

jan-wassenberg commented on meson_cpu/x86/meson.build:
```diff
@@ -1,228 +1,106 @@
+)
+X86_V3 = mod_features.new(
+  'X86_V3', 10, implies: X86_V2, 
+  args: ['-mavx', '-mavx2', '-mfma', '-mbmi', '-mbmi2', '-mlzcnt', '-mf16c', '-mmovbe'],
```
I agree HWY_DISABLE_PCLMUL_AES is sufficient for avoiding the need for CLMUL and AES. If defined, `CLMul*` and `AESRound` must not be called, which you don't, so all good.

r-devulap reviewed (approved):
LGTM.

h-vetinari reviewed (commented):
There are comments about fixes needing to be backported. Has this happened?

rgommers commented:
Just a quick note: we _really_ don't care about performance on 32-bit Windows, we don't even bundle OpenBLAS so linalg operations are very slow, and we don't get any complaints. It just needs to work with minimal maintenance overhead.
